File: backend/app/services/form_error_service.py
"""
Form Error Service for handling form validation and error messages
Provides consistent error handling across all forms in the application
"""
from typing import Dict, Any, Optional
from fastapi import HTTPException


class FormErrorService:
    """Service for handling form validation and error messages"""
    
    # Standard error messages
    ERROR_MESSAGES = {
        'required_field': '{field_name} is required',
        'invalid_credentials': 'Invalid username or password',
        'invalid_format': 'Please enter a valid {field_type}',
        'server_error': 'Server error. Please try again later.',
        'role_mismatch': 'Selected role combination is not valid',
        'password_required': 'Password is required',
        'username_required': 'Username is required',
        'main_role_required': 'Please select a main role',
        'sub_role_required': 'Please select a sub role',
        'specialization_required': 'Please select a specialization'
    }
    
    def validate_login_form(self, form_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Validate login form data
        
        Args:
            form_data: Dictionary containing username and password
            
        Returns:
            Dictionary with validation results
        """
        errors = {}
        
        # Check username
        if not form_data.get('username') or str(form_data['username']).strip() == '':
            errors['username'] = self.ERROR_MESSAGES['username_required']
        
        # Check password
        if not form_data.get('password') or str(form_data['password']).strip() == '':
            errors['password'] = self.ERROR_MESSAGES['password_required']
        
        return {
            'is_valid': len(errors) == 0,
            'errors': errors
        }
    
    def validate_role_selection(self, role_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Validate role selection form data
        
        Args:
            role_data: Dictionary containing role selection data
            
        Returns:
            Dictionary with validation results
        """
        errors = {}
        
        # Check main role
        if not role_data.get('main_role') or str(role_data['main_role']).strip() == '':
            errors['main_role'] = self.ERROR_MESSAGES['main_role_required']
        
        # Check sub role
        if not role_data.get('sub_role') or str(role_data['sub_role']).strip() == '':
            errors['sub_role'] = self.ERROR_MESSAGES['sub_role_required']
        
        # Check specialization
        if not role_data.get('specialization') or str(role_data['specialization']).strip() == '':
            errors['specialization'] = self.ERROR_MESSAGES['specialization_required']
        
        return {
            'is_valid': len(errors) == 0,
            'errors': errors
        }
    
    def handle_server_error(self, error: Any) -> str:
        """
        Convert server errors to user-friendly messages
        
        Args:
            error: Error object or HTTP status code
            
        Returns:
            User-friendly error message
        """
        if hasattr(error, 'status_code'):
            status_code = error.status_code
        elif hasattr(error, 'status'):
            status_code = error.status
        elif isinstance(error, int):
            status_code = error
        else:
            status_code = 500
        
        if status_code == 401:
            return self.ERROR_MESSAGES['invalid_credentials']
        elif status_code == 400:
            if hasattr(error, 'data') and error.data and error.data.get('message'):
                return error.data['message']
            return 'Invalid input data'
        elif status_code >= 500:
            return self.ERROR_MESSAGES['server_error']
        else:
            return 'An unexpected error occurred. Please try again.'
    
    def validate_field_format(self, field_name: str, field_value: Any, field_type: str) -> Optional[str]:
        """
        Validate field format
        
        Args:
            field_name: Name of the field
            field_value: Value to validate
            field_type: Expected type (email, phone, etc.)
            
        Returns:
            Error message if validation fails, None if valid
        """
        if field_type == 'email':
            import re
            email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
            if not re.match(email_pattern, str(field_value)):
                return f'Please enter a valid email address'
        
        return None
    
    def get_field_specific_error(self, field_name: str, error_type: str) -> str:
        """
        Get field-specific error message
        
        Args:
            field_name: Name of the field
            error_type: Type of error
            
        Returns:
            Formatted error message
        """
        if error_type == 'required':
            return self.ERROR_MESSAGES['required_field'].format(field_name=field_name.replace('_', ' ').title())
        elif error_type == 'invalid_format':
            return self.ERROR_MESSAGES['invalid_format'].format(field_type=field_name.replace('_', ' '))
        else:
            return f'Invalid {field_name.replace("_", " ")}'

File: frontend/src/js/services/form-error.service.js
/**
 * Form Error Service - Comprehensive form validation and error handling
 * Provides validation methods for all forms with specific error messages
 */
(function() {
    'use strict';

    angular
        .module('interviewPrepApp')
        .service('FormErrorService', FormErrorService);

    FormErrorService.$inject = ['$log'];

    function FormErrorService($log) {
        var service = {
            // Login form validation
            validateLoginForm: validateLoginForm,
            
            // Role selection validation
            validateRoleSelection: validateRoleSelection,
            
            // Server error handling
            handleServerError: handleServerError,
            
            // Generic form validation
            validateRequiredFields: validateRequiredFields,
            validateEmailFormat: validateEmailFormat,
            validatePasswordStrength: validatePasswordStrength,
            
            // Error message utilities
            getFieldErrorMessage: getFieldErrorMessage,
            formatValidationErrors: formatValidationErrors
        };

        return service;

        // ==================== Login Form Validation ====================

        /**
         * Validate login form data
         * @param {Object} formData - Login form data with email/username and password
         * @returns {Object} Validation result with isValid flag and errors object
         */
        function validateLoginForm(formData) {
            var errors = {};
            
            if (!formData) {
                return {
                    isValid: false,
                    errors: { form: 'Form data is required' }
                };
            }

            // Validate email/username field
            var emailField = formData.email || formData.username;
            if (!emailField || emailField.trim() === '') {
                errors.email = 'Email is required';
                errors.username = 'Username is required';
            } else if (formData.email && !validateEmailFormat(emailField)) {
                errors.email = 'Please enter a valid email address';
            }

            // Validate password field
            if (!formData.password || formData.password.trim() === '') {
                errors.password = 'Password is required';
            }

            var result = {
                isValid: Object.keys(errors).length === 0,
                errors: errors
            };

            $log.debug('FormErrorService: Login form validation result', result);
            return result;
        }

        // ==================== Role Selection Validation ====================

        /**
         * Validate role selection form data
         * @param {Object} roleData - Role selection data
         * @returns {Object} Validation result with isValid flag and errors object
         */
        function validateRoleSelection(roleData) {
            var errors = {};
            
            if (!roleData) {
                return {
                    isValid: false,
                    errors: { form: 'Role selection data is required' }
                };
            }

            // Validate main role
            if (!roleData.mainRole || roleData.mainRole.trim() === '') {
                errors.mainRole = 'Please select a main role';
            }

            // Validate sub role
            if (!roleData.subRole || roleData.subRole.trim() === '') {
                errors.subRole = 'Please select a sub role';
            }

            // Validate specialization (optional but if provided, should not be empty)
            if (roleData.specialization && roleData.specialization.trim() === '') {
                errors.specialization = 'Please select a valid specialization or leave empty';
            }

            // Validate role combination compatibility
            if (roleData.mainRole && roleData.subRole) {
                var compatibilityError = validateRoleCompatibility(roleData.mainRole, roleData.subRole);
                if (compatibilityError) {
                    errors.roleCompatibility = compatibilityError;
                }
            }

            // Validate tech stack selection (should not exceed 3 items)
            if (roleData.techStack && Array.isArray(roleData.techStack)) {
                if (roleData.techStack.length > 3) {
                    errors.techStack = 'Please select no more than 3 tech stack items';
                }
            }

            var result = {
                isValid: Object.keys(errors).length === 0,
                errors: errors
            };

            $log.debug('FormErrorService: Role selection validation result', result);
            return result;
        }

        /**
         * Validate role combination compatibility
         * @param {string} mainRole - Main role selection
         * @param {string} subRole - Sub role selection
         * @returns {string|null} Error message if incompatible, null if valid
         */
        function validateRoleCompatibility(mainRole, subRole) {
            // Define incompatible role combinations
            var incompatibleCombinations = {
                'Student': ['Senior Developer', 'Team Lead', 'Engineering Manager', 'CTO'],
                'Entry Level': ['Senior Developer', 'Team Lead', 'Engineering Manager', 'CTO'],
                'Junior': ['Senior Developer', 'Team Lead', 'Engineering Manager']
            };

            if (incompatibleCombinations[mainRole] && 
                incompatibleCombinations[mainRole].includes(subRole)) {
                return 'Selected role combination is not valid. ' + mainRole + ' cannot be ' + subRole;
            }

            return null;
        }

        // ==================== Server Error Handling ====================

        /**
         * Handle server errors and convert to user-friendly messages
         * @param {Object} error - Server error object
         * @returns {string} User-friendly error message
         */
        function handleServerError(error) {
            if (!error) {
                return 'An unexpected error occurred. Please try again.';
            }

            var status = error.status || 0;
            var errorData = error.data || {};

            // Handle specific HTTP status codes
            switch (status) {
                case 400:
                    return handleBadRequestError(errorData);
                case 401:
                    return 'Invalid username or password. Please check your credentials and try again.';
                case 403:
                    return 'You don\'t have permission to perform this action. Please contact support.';
                case 404:
                    return 'The requested resource was not found. Please check the URL and try again.';
                case 409:
                    return 'This action conflicts with existing data. Please refresh and try again.';
                case 422:
                    return handleValidationError(errorData);
                case 429:
                    return 'Too many requests. Please wait a few minutes before trying again.';
                case 500:
                    return 'Server error. Please try again later or contact support if the problem persists.';
                case 502:
                case 503:
                case 504:
                    return 'Service temporarily unavailable. Please try again in a few minutes.';
                case 0:
                    return 'Network error. Please check your internet connection and try again.';
                default:
                    return 'An unexpected error occurred. Please try again or contact support.';
            }
        }

        /**
         * Handle 400 Bad Request errors
         * @param {Object} errorData - Error data from server
         * @returns {string} User-friendly error message
         */
        function handleBadRequestError(errorData) {
            if (errorData.message) {
                return errorData.message;
            }
            
            if (errorData.detail) {
                if (typeof errorData.detail === 'string') {
                    return errorData.detail;
                }
                if (errorData.detail.message) {
                    return errorData.detail.message;
                }
            }

            return 'Invalid request. Please check your input and try again.';
        }

        /**
         * Handle 422 Validation errors
         * @param {Object} errorData - Error data from server
         * @returns {string} User-friendly error message
         */
        function handleValidationError(errorData) {
            if (errorData.detail && Array.isArray(errorData.detail)) {
                // FastAPI validation error format
                var messages = errorData.detail.map(function(err) {
                    return err.msg || 'Validation error';
                });
                return 'Validation failed: ' + messages.join(', ');
            }

            if (errorData.message) {
                return errorData.message;
            }

            return 'Please check your input data and try again.';
        }

        // ==================== Generic Validation Methods ====================

        /**
         * Validate required fields in a form
         * @param {Object} formData - Form data object
         * @param {Array} requiredFields - Array of required field names
         * @returns {Object} Validation result
         */
        function validateRequiredFields(formData, requiredFields) {
            var errors = {};

            if (!formData || !requiredFields) {
                return {
                    isValid: false,
                    errors: { form: 'Invalid form data or field configuration' }
                };
            }

            requiredFields.forEach(function(fieldName) {
                var fieldValue = formData[fieldName];
                if (!fieldValue || (typeof fieldValue === 'string' && fieldValue.trim() === '')) {
                    errors[fieldName] = getFieldErrorMessage(fieldName, 'required');
                }
            });

            return {
                isValid: Object.keys(errors).length === 0,
                errors: errors
            };
        }

        /**
         * Validate email format
         * @param {string} email - Email address to validate
         * @returns {boolean} True if valid email format
         */
        function validateEmailFormat(email) {
            if (!email || typeof email !== 'string') {
                return false;
            }

            var emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
            return emailRegex.test(email.trim());
        }

        /**
         * Validate password strength
         * @param {string} password - Password to validate
         * @returns {Object} Validation result with strength info
         */
        function validatePasswordStrength(password) {
            var result = {
                isValid: true,
                strength: 'weak',
                errors: [],
                suggestions: []
            };

            if (!password || typeof password !== 'string') {
                result.isValid = false;
                result.errors.push('Password is required');
                return result;
            }

            var length = password.length;
            var hasLower = /[a-z]/.test(password);
            var hasUpper = /[A-Z]/.test(password);
            var hasNumber = /\d/.test(password);
            var hasSpecial = /[!@#$%^&*(),.?":{}|<>]/.test(password);

            // Check minimum requirements
            if (length < 8) {
                result.isValid = false;
                result.errors.push('Password must be at least 8 characters long');
            }

            // Determine strength
            var strengthScore = 0;
            if (length >= 8) strengthScore++;
            if (hasLower) strengthScore++;
            if (hasUpper) strengthScore++;
            if (hasNumber) strengthScore++;
            if (hasSpecial) strengthScore++;

            if (strengthScore >= 4) {
                result.strength = 'strong';
            } else if (strengthScore >= 3) {
                result.strength = 'medium';
            } else {
                result.strength = 'weak';
            }

            // Add suggestions for improvement
            if (!hasLower) result.suggestions.push('Add lowercase letters');
            if (!hasUpper) result.suggestions.push('Add uppercase letters');
            if (!hasNumber) result.suggestions.push('Add numbers');
            if (!hasSpecial) result.suggestions.push('Add special characters');
            if (length < 12) result.suggestions.push('Use at least 12 characters for better security');

            return result;
        }

        // ==================== Error Message Utilities ====================

        /**
         * Get specific error message for a field and error type
         * @param {string} fieldName - Name of the field
         * @param {string} errorType - Type of error (required, invalid, etc.)
         * @returns {string} Formatted error message
         */
        function getFieldErrorMessage(fieldName, errorType) {
            var fieldLabels = {
                'email': 'Email',
                'username': 'Username',
                'password': 'Password',
                'confirmPassword': 'Confirm Password',
                'mainRole': 'Main Role',
                'subRole': 'Sub Role',
                'specialization': 'Specialization',
                'techStack': 'Tech Stack',
                'firstName': 'First Name',
                'lastName': 'Last Name',
                'phone': 'Phone Number'
            };

            var errorMessages = {
                'required': '{field} is required',
                'invalid': 'Please enter a valid {field}',
                'format': 'Please enter {field} in the correct format',
                'mismatch': '{field} does not match',
                'exists': '{field} already exists',
                'length': '{field} must meet length requirements'
            };

            var fieldLabel = fieldLabels[fieldName] || fieldName;
            var messageTemplate = errorMessages[errorType] || 'Invalid {field}';

            return messageTemplate.replace('{field}', fieldLabel);
        }

        /**
         * Format validation errors for display
         * @param {Object} errors - Errors object from validation
         * @returns {Array} Array of formatted error messages
         */
        function formatValidationErrors(errors) {
            if (!errors || typeof errors !== 'object') {
                return [];
            }

            return Object.keys(errors).map(function(fieldName) {
                return {
                    field: fieldName,
                    message: errors[fieldName]
                };
            });
        }
    }
})();